pkg/event-sidecar/cmd.go

```diff
@@ -0,0 +1,143 @@
+///////////////////////////////////////////////////////////////////////
+// Copyright (c) 2017 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+
+package eventsidecar
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/vmware/dispatch/pkg/event-sidecar/listener"
+	"github.com/vmware/dispatch/pkg/events"
+	"github.com/vmware/dispatch/pkg/events/parser"
+	"github.com/vmware/dispatch/pkg/events/transport"
+	"github.com/vmware/dispatch/pkg/events/validator"
+)
+
+// NO TESTS
+
+type sidecarConfig struct {
+	ListenerProtocol string `mapstructure:"listener-protocol"`
+	ListenerPort     int    `mapstructure:"listener-port"`
+	ListenerPipe     string `mapstructure:"listener-pipe"`
+	Transport        string `mapstructure:"transport"`
+	RabbitMQURL      string `mapstructure:"rabbitmq-url"`
+	Tenant           string `mapstructure:"tenant"`
+	TracerURL        string `mapstructure:"tracer-url"`
+	Debug            bool   `mapstructure:"debug"`
+}
+
+var sidecarCfg sidecarConfig
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+	viper.SetEnvPrefix("dispatch")
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+}
+
+// NewCmd creates a top-level command for event driver sidecar
+func NewCmd(in io.Reader, out, errOut io.Writer) *cobra.Command {
+	cmds := &cobra.Command{
+		Use: "event-driver-sidecar",
+		Run: sidecarCmd,
+	}
+
+	cobra.OnInitialize(func() {
+		if err := viper.Unmarshal(&sidecarCfg); err != nil {
+			log.Fatalf("Error while unmarshalling configuration: %s", err)
+		}
+		if sidecarCfg.Debug {
+			log.SetLevel(log.DebugLevel)
+		} else {
+			log.SetLevel(log.InfoLevel)
+		}
+	})
+
+	cmds.PersistentFlags().String("listener-protocol", "http", "Protocol to be used by listener. One of [http, grpc, pipe] ($DISPATCH_LISTENER_PROTOCOL)")
```

Comment:
So is it one sidecar per driver?  Could we have multiple drivers writing to the same sidecar?  Could the sidecar listen on multiple protocols?

Reply:
You got me thinking with that one. Making sidecar to listen using all protocols at once sounds reasonable, that would solve the "mode" flag in driver type.

The idea behind sidecar is that each driver gets its own. Each driver instance is then a pod, and can communicate over local TCP/IP stack (or files on shared volume, or IPC).
Now, updating single pod with more driver containers does not sound reasonable, as it would have to recreated every time (adding new containers to running pod is not allowed per: https://github.com/kubernetes/kubernetes/issues/37838). 
If we used separate pods, we would lose the ability to use localhost or shared volume. We could only use networking stack, and we would get into service discovery of where the sidecar/service lives. Also, sidecar encapsulates privileges/message routing, so that event driver doesn't have to. Because we have one sidecar per event driver, we can configure the routing/privileges at the time of driver pod deployment and not worry about it later.

However, I'm seeing some advantages from performance perspective. Maybe we could investigate having one sidecar (well, not sidecar anymore, let's call it event router perhaps) per tenant. This would ensure the routing/privileges stay the same.